### PR TITLE
Replace cuco stream ref with `cuda::stream_ref`

### LINKED ISSUE #331: Switch to cuda::stream_ref
`cuco::stream_ref` is a WAR and should be replaced with `cuda::stream_ref`.

`cuda::stream_ref` will be available in libcu++ v2.1.0, and we should get that version soon: rapidsai/rapids-cmake#399.

_Originally posted by @sleeepyjack in https://github.com/NVIDIA/cuCollections/issues/296#issuecomment-1522065120_

### PR DESCRIPTION
**Pull request by PointKernel**
Closes #331 

Fixes #413

This PR eliminates the cuco custom stream ref implementations by using `cuda::stream_ref`.

### CODE REVIEW

File: include/cuco/detail/hyperloglog/hyperloglog_ref.cuh [hunk 1]
```diff
@@ -122,21 +122,21 @@ class hyperloglog_ref {
...
    * @brief Asynchronously resets the estimator, i.e., clears the current count estimate.
    *
    * @param stream CUDA stream this operation is executed in
    */
-  __host__ constexpr void clear_async(cuco::cuda_stream_ref stream) noexcept
+  __host__ constexpr void clear_async(cuda::stream_ref stream) noexcept
   {
     auto constexpr block_size = 1024;
-    cuco::hyperloglog_ns::detail::clear<<<1, block_size, 0, stream>>>(*this);
+    cuco::hyperloglog_ns::detail::clear<<<1, block_size, 0, stream.get()>>>(*this);
```
Comment (sleeepyjack) [inline include/cuco/detail/hyperloglog/hyperloglog_ref.cuh:139]: Is the get() required? No implicit conversion?

File: include/cuco/detail/hyperloglog/hyperloglog_ref.cuh [hunk 2]
```diff
@@ -122,21 +122,21 @@ class hyperloglog_ref {
    *
    * @param stream CUDA stream this operation is executed in
    */
-  __host__ constexpr void clear(cuco::cuda_stream_ref stream)
+  __host__ constexpr void clear(cuda::stream_ref stream)
   {
     this->clear_async(stream);
-    stream.synchronize();
+    stream.wait();
```
Comment (sleeepyjack) [inline include/cuco/detail/hyperloglog/hyperloglog_ref.cuh:128]: Just to make sure: wait() also throws the last CUDA error?

**Review by sleeepyjack — APPROVED** [review]
Nice cleanup!
LGTM apart from some minor cuda::stream_ref API questions to make sure it behaves equivalent to our stream_ref class.

File: include/cuco/detail/hyperloglog/hyperloglog_ref.cuh (hunk 2 above)
Comment (PointKernel) [reply include/cuco/detail/hyperloglog/hyperloglog_ref.cuh:128]: Yes, I've checked the code it throws
https://github.com/NVIDIA/cccl/blob/3104dd086d6b22be1c9dd45b64550eb70635d610/libcudacxx/include/cuda/stream_ref#L144

File: include/cuco/detail/hyperloglog/hyperloglog_ref.cuh (hunk 1 above)
Comment (PointKernel) [reply include/cuco/detail/hyperloglog/hyperloglog_ref.cuh:139]: Yes, it is required.

**Pull request merged**